packages/cli/src/modules/mcp/tools/execute-workflow.tool.ts

```diff
@@ -0,0 +1,275 @@
+	result: z.any().optional().describe('Workflow execution result data'),
+	error: z.unknown().optional(),
+} satisfies z.ZodRawShape;
+
+export const createExecuteWorkflowTool = (
+	user: User,
+	workflowFinderService: WorkflowFinderService,
+	activeExecutions: ActiveExecutions,
+	workflowRunner: WorkflowRunner,
+	telemetry: Telemetry,
+): ToolDefinition<typeof inputSchema.shape> => ({
+	name: 'execute_workflow',
+	config: {
+		description: 'Execute a workflow by id',
+		inputSchema: inputSchema.shape,
+		outputSchema,
+	},
+	handler: async ({ workflowId, inputs }) => {
+		const telemetryPayload: UserCalledMCPToolEventPayload = {
+			user_id: user.id,
+			tool_name: 'execute_workflow',
+			parameters: { workflowId, inputs },
+		};
+		try {
+			const output = await executeWorkflow(
+				user,
+				workflowFinderService,
+				activeExecutions,
+				workflowRunner,
+				workflowId,
+				inputs,
+			);
+
+			telemetryPayload.results = {
+				success: output.success,
+				data: {
+					executionId: output.executionId,
+				},
+			};
+			telemetry.track(USER_CALLED_MCP_TOOL_EVENT, telemetryPayload);
+
+			return {
+				content: [{ type: 'text', text: JSON.stringify(output) }],
+				structuredContent: output,
+			};
+		} catch (error) {
+			const payload: ExecuteWorkflowOutput = {
+				success: false,
+				executionId: null,
+				error,
+			};
+
+			telemetryPayload.results = {
+				success: false,
+				error: error instanceof Error ? error.message : String(error),
+			};
+			telemetry.track(USER_CALLED_MCP_TOOL_EVENT, telemetryPayload);
+
+			return {
+				content: [{ type: 'text', text: JSON.stringify(payload) }],
+				structuredContent: payload,
+			};
+		}
+	},
+});
+
+/**
+ * Executes a workflow for the given user with provided inputs.
+ * In order to "synchronously" execute the workflow,
+ * it is mapping mcp tool inputs to trigger node pin data and starting execution from there.
+ * LIMITATION: Does not properly support workflows with multiple triggers.
+ */
+export const executeWorkflow = async (
+	user: User,
+	workflowFinderService: WorkflowFinderService,
+	activeExecutions: ActiveExecutions,
+	workflowRunner: WorkflowRunner,
+	workflowId: string,
+	inputs?: z.infer<typeof inputSchema>['inputs'],
+): Promise<ExecuteWorkflowOutput> => {
+	const workflow = await workflowFinderService.findWorkflowForUser(workflowId, user, [
+		'workflow:read',
+	]);
+
+	if (!workflow || workflow.isArchived || !workflow.settings?.availableInMCP) {
+		throw new UserError('Workflow not found');
+	}
+
+	const canExecuteWorkflow = isWorkflowEligibleForMCPAccess(workflow);
+
+	if (!canExecuteWorkflow) {
+		throw new UserError(
+			`Only workflows with the following trigger nodes can be executed: ${Object.values(SUPPORTED_MCP_TRIGGERS).join(', ')}.`,
+		);
+	}
+
+	const triggerNode = findMcpWorkflowStart(workflow.nodes);
+
+	if (!triggerNode) {
+		throw new UserError('No supported trigger node found to start the workflow execution.');
+	}
```

Comment:
We're looping again, maybe `isWorkflowEligibleForMCPAccess` should already return the first eligible trigger.